RecoJets/JetProducers/plugins/FixedGridRhoProducerFastjetFromRecHit.cc

```diff
@@ -0,0 +1,180 @@
+
+  if (skipHCAL_ && skipECAL_) {
+    throw cms::Exception("FixedGridRhoProducerFastjetFromRecHit")
+        << "skipHCAL and skipECAL both can't be True. Make at least one of them False.";
+  }
+
+  if (!skipHCAL_) {
+    for (const auto &hit : iEvent.get(hbheRecHitsTag_)) {
+      if (passedHcalNoiseCut(hit)) {
+        TLorentzVector hitp4(0, 0, 0, 0);
+        getHitP4(hit.id(), hit.energy(), hitp4, iSetup.getData(caloGeometryToken_));
+        inputs.push_back(fastjet::PseudoJet(hitp4.Px(), hitp4.Py(), hitp4.Pz(), hitp4.E()));
+      }
+    }
+  }
+
+  if (!skipECAL_) {
+    for (const auto &hit : iEvent.get(ebRecHitsTag_)) {
+      if (passedEcalNoiseCut(hit, &thresholds)) {
+        TLorentzVector hitp4(0, 0, 0, 0);
+        getHitP4(hit.id(), hit.energy(), hitp4, iSetup.getData(caloGeometryToken_));
+        inputs.push_back(fastjet::PseudoJet(hitp4.Px(), hitp4.Py(), hitp4.Pz(), hitp4.E()));
+      }
+    }
+
+    for (const auto &hit : iEvent.get(eeRecHitsTag_)) {
+      if (passedEcalNoiseCut(hit, &thresholds)) {
+        TLorentzVector hitp4(0, 0, 0, 0);
+        getHitP4(hit.id(), hit.energy(), hitp4, iSetup.getData(caloGeometryToken_));
+        inputs.push_back(fastjet::PseudoJet(hitp4.Px(), hitp4.Py(), hitp4.Pz(), hitp4.E()));
+      }
+    }
+  }
+
+  bge_.set_particles(inputs);
+  iEvent.put(std::make_unique<double>(bge_.rho()));
+}
+
+void FixedGridRhoProducerFastjetFromRecHit::getHitP4(const DetId &detId,
+                                                     float hitE,
+                                                     TLorentzVector &hitp4,
+                                                     const CaloGeometry &caloGeometry) {
+  const CaloSubdetectorGeometry *subDetGeom = caloGeometry.getSubdetectorGeometry(detId);
+  std::shared_ptr<const CaloCellGeometry> cellGeom =
+      subDetGeom != nullptr ? subDetGeom->getGeometry(detId) : std::shared_ptr<const CaloCellGeometry>();
+  if (cellGeom != nullptr) {
+    const auto &gpPos = cellGeom->repPos();
+    double thispt = hitE / cosh(gpPos.eta());
+    double thispx = thispt * cos(gpPos.phi());
+    double thispy = thispt * sin(gpPos.phi());
+    double thispz = thispt * sinh(gpPos.eta());
+    hitp4.SetPxPyPzE(thispx, thispy, thispz, hitE);
+  } else {
+    if (detId.rawId() != 0)
+      edm::LogInfo("FixedGridRhoProducerFastjetFromRecHit")
+          << "Warning : Geometry not found for a calo hit, setting p4 as (0,0,0,0)" << std::endl;
+    hitp4.SetPxPyPzE(0, 0, 0, 0);
+  }
+}
+
+//HCAL noise cleaning cuts.
+bool FixedGridRhoProducerFastjetFromRecHit::passedHcalNoiseCut(const HBHERecHit &hit) {
+  bool passed = false;
+  const HcalDetId thisDetId(hit.detid());
+  const int thisDepth = thisDetId.depth();
+  if ((thisDetId.subdet() == HcalBarrel) && (hit.energy() > eThresHB_[thisDepth - 1]))
+    passed = true;
+  else if ((thisDetId.subdet() == HcalEndcap) && (hit.energy() > eThresHE_[thisDepth - 1]))
+    passed = true;
+  return passed;
```

Comment:
```suggestion
  const auto thisDetId = hit.detid();
  const auto thisDepth = thisDetId.depth();
  if (thisDetId.subdet() == HcalBarrel && hit.energy() > eThresHB_[thisDepth-1])
    return true;
  else if (thisDetId.subdet() == HcalEndcap && hit.energy() > eThresHE_[thisDepth-1])
    return true;
  return false;
```

Reply:
I have now changed it as you suggested; just one minor thing is that,
I have done 
`const HcalDetId thisDetId = hit.detid();` 
instead of
`const auto thisDetId = hit.detid();`
Unless I explicitly specify that `thisDetId` is of type `HcalDetId`, the compiler seem to think that `thisDetId` is of type `DetId`. 
`HcalDetId` inherits from `DetId`, but `depth()` which is accessed in next line, is available only in `HcalDetId` and not in `DetId`. So it raises an error.

Reply:
Sure, no problem. Just FYI, it looks like you could use [`HBHERecHit::id()`](https://github.com/cms-sw/cmssw/blob/c44be9acb12672c530d1ed454901e1435f60a2cc/DataFormats/HcalRecHit/interface/HBHERecHit.h#L41) (in any case, this is RECO, so not my domain).

Reply:
right, `const auto thisDetId = hit.id();` works fine; so moving to that.. 

and all other comments are also addressed now. Will push the updated version soon.